politeiawww/cmd/politeiavoteraudit/politeiavoteraudit.go

```diff
@@ -0,0 +1,724 @@
+	if verbose {
+		fmt.Printf("  Parsing: %v\n", filepath.Base(filename))
+	}
+
+	f, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	first := true
+	d := json.NewDecoder(f)
+	for {
+		// Read timestamp
+		var timestamp jsontypes.Timestamp
+		if err := d.Decode(&timestamp); err == io.EOF {
+			if first {
+				return fmt.Errorf("expected timestamp in " +
+					"failed log")
+			}
+			break
+		} else if err != nil {
+			return err
+		}
+
+		// Read Ballot
+		var b v1.Ballot
+		if err := d.Decode(&b); err == io.EOF {
+			return fmt.Errorf("no ballot in failed log")
+		} else if err != nil {
+			return err
+		}
+
+		// Read ErrRetry
+		var e HackError
+		if err := d.Decode(&e); err == io.EOF {
+			return fmt.Errorf("no errretry in failed log")
+		} else if err != nil {
+			return err
+		}
+
+		t, err := time.Parse(time.StampNano, timestamp.Time)
+		if err != nil {
+			return err
+		}
+		m[t.Unix()] = BallotError{
+			Ballot: b,
+			Error:  e,
+		}
+
+		//fmt.Printf("%v%v\n", spew.Sdump(b), spew.Sdump(e))
+		first = false
+	}
+
+	return nil
+}
+
+func findSuccess(ticket string, m map[int64]jsontypes.BallotResult) bool {
+	for _, v := range m {
+		if v.Ticket != ticket {
+			continue
+		}
+		return true
+	}
+	return false
+}
+
+func countSuccess(proposal, ticket string, m map[int64]jsontypes.BallotResult) int {
+	var count int
+	for _, v := range m {
+		if v.Ticket != ticket {
+			continue
+		}
+		count++
+	}
+	return count
+}
+
+func countFailed(proposal, ticket string, m map[int64]BallotError) int {
+	var count int
+	for _, v := range m {
+		for _, vv := range v.Ballot.Votes {
+			if !(vv.Token == proposal && vv.Ticket == ticket) {
+				continue
+			}
+			count++
+		}
+	}
+	return count
+}
+
+func findWork(ticket string, m map[int64][]jsontypes.VoteInterval) bool {
+	for _, v := range m {
+		for _, vv := range v {
+			if ticket == vv.Vote.Ticket {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func auditWork(verbose bool, work map[int64][]jsontypes.VoteInterval, success map[int64]jsontypes.BallotResult, failed map[int64]BallotError) (int, error) {
+	if verbose {
+		fmt.Printf("  Audit: work\n")
+	}
+
+	// Read all work items and ensure they made it to one of the success
+	// files.
+	var errorCount, successCount, totalRecords int
+	for _, v := range work {
+		totalRecords += len(v)
+		for _, vv := range v {
+			if verbose {
+				fmt.Printf("    Checking ticket: %v\n",
+					vv.Vote.Ticket)
+			}
+
+			// Troll sucess
+			s := countSuccess(vv.Vote.Token, vv.Vote.Ticket,
+				success)
+			if verbose {
+				fmt.Printf("    Success           : %v\n", s)
+			}
+			if s != 1 {
+				errorCount++
+				fmt.Printf("    Unexpected success: "+
+					"want %v got %v\n", 1, s)
+			}
+
+			// Troll failed
+			f := countFailed(vv.Vote.Token, vv.Vote.Ticket, failed)
+			if verbose {
+				fmt.Printf("    Retries           : %v\n", f)
+			}
+
+			if s == 0 && f <= 0 {
+				errorCount++
+				fmt.Printf("    Unexpected failed : "+
```

Comment:
These error messages are pretty confusing.  These errors are what get printed for a ticket that exists in the work journal, but hasn't voted at all yet:
```
    Unexpected success: want 1 got 0
    Unexpected failed : want > 0 got 0
```